src/memoization/memoization.jl

```diff
@@ -0,0 +1,48 @@
+mutable struct MemoNLP <: AbstractNLPModel
+  model :: AbstractNLPModel
+  meta :: NLPModelMeta
+  counters :: Counters
+  min_elm :: Int
+  max_elm :: Int
+  memo :: Dict{AbstractVector, AbstractFloat}
+  freq :: Dict{AbstractVector, Int}
+end
+
+function MemoNLP(model :: AbstractNLPModel, max_elm :: Int, min_elm :: Int)
+  max_elm != 0 || return model
+
+  x0 = model.meta.x0
+  ncon = model.meta.ncon
+  lvar = model.meta.lvar
+  uvar = model.meta.uvar
+  lcon = model.meta.lcon
+  ucon = model.meta.ucon
+  T = eltype(model.meta.x0)
+  memo = Dict{AbstractVector{T}, T}()
+  freq = Dict{T, Int}()
+  meta = NLPModelMeta(model.meta.nvar, x0=x0, ncon = ncon, lvar=lvar, uvar=uvar, lcon=lcon, ucon=ucon, nnzj=0, nnzh=0)
+
+  return MemoNLP(model, meta, Counters(), min_elm, max_elm, memo, freq)
+end
+
+function clean_memo!(mnlp :: MemoNLP)
+  for i in sort(collect(mnlp.freq), by=x->x[2])[1:mnlp.min_elm]
+    delete!(mnlp.memo, i[1])
+    delete!(mnlp.freq, i[1])
+  end
+
+  return mnlp
+end
+
+function NLPModels.obj(mnlp :: MemoNLP, x :: AbstractVector)
+  xt = copy(x)
+  if haskey(mnlp.memo, xt)
```

Comment:
Also had to use a ```copy(x)``` due to pointer behaviour. However, the Dict keeps breaking at some tests, so i'll have to take a deeper look into it.